### PR TITLE
consensus: accept Dusk-signed Emergency Block

### PR DESCRIPTION
**Pull request by fed-franz**
Partially implements https://github.com/dusk-network/rusk/issues/1234

New features:
It enables the acceptance of an Emergency Block (EB) with the following properties:
    -  `iteration = u8::MAX`
    - `timestamp > prev_block.timestamp + MIN_EMERGENCY_BLOCK_TIME`, where `MIN_EMERGENCY_BLOCK_TIME` is  of  `MAX_STEP_TIMEOUT x 3`  (`MAX_ITER_TIMEOUT`) times CONSENSUS_MAX_ITER + 1. This value aims at ensure enough time is given to provisioners to produce a block before an Emergency Block can be accepted.
    - signed by Dusk (`DUSK_CONSENSUS_KEY`)

No Attestation is needed to accept an EB. 

For what concerns finality, the Emergency Block is consider as being produced at iteration CONSENSUS_MAX_ITER (i.e., the one after the last consensus iteration)

The EB is repropagated upon acceptance (unlike other Block messages)

---

Refactors:
 - on_block_event: reduces indentation and improves code clarity by splitting `attach_att_if_needed` into `is_block_attested` and `attach_blk_att`
 - `execute_checks`: change `disable_att_check` into `check_attestation`
 - rusk/node: remove duplicate Dusk key loading
 - `generate_block`: move `seed` computation and "block generation elapsed-time log" from `generate_candidate_message` to `generate_block`; take `voters` from `ru`
 - remove iteration check from message deserialization (this also allows to have a standalone definition of `CONSENSUS_MAX_ITER`, instead of tieing it to `MESSAGE_MAX_ITER`

### CODE REVIEW

File: node/src/chain/acceptor.rs [hunk 1]
```diff
@@ -48,6 +48,8 @@ use crate::database::rocksdb::{
 use crate::database::{self, ConsensusStorage, Ledger, Mempool, Metadata};
 use crate::{vm, Message, Network};
 
+use rusk_recovery_tools::state::DUSK_CONSENSUS_KEY;
```
Comment (herr-seppia) [inline node/src/chain/acceptor.rs:51]: Do you mind to "pub use" it in `node/src/lib.rs` and use it here?
So to remove the additional declaration in `rusk/src/lib/node/rusk.rs`?

File: node/src/chain/acceptor.rs [hunk 2]
```diff
@@ -1337,13 +1339,29 @@ pub(crate) async fn verify_block_header<DB: database::DB>(
...
+    // Set the expected generator to the one extracted by Deterministic
+    // Sortition, or, in case of Emergency Block, to the Dusk Consensus Key
+    let (expected_generator, check_att) =
+        if is_emergency_block(header.iteration) {
+            let dusk_key = PublicKey::new(*DUSK_CONSENSUS_KEY);
+            let dusk_key_bytes = dusk_key.bytes();
+
+            // We disable the Attestation check since it's not needed to accept
+            // an Emergency Block
+            (*dusk_key_bytes, false)
```
Comment (herr-seppia) [inline node/src/chain/acceptor.rs:1351]: Do you need to add an exception when accepting the next block in order to not verify previous (missing) attestation?

File: node/src/chain/fsm.rs [hunk 3]
```diff
@@ -187,134 +190,139 @@ impl<N: Network, DB: database::DB, VM: vm::VMExecution> SimpleFSM<N, DB, VM> {
...
-                                    event = "revert failed",
-                                    err = format!("{:?}", err)
-                                );
-                            }
-                        }
-                        Err(e) => {
-                            error!(
                                 event = "revert failed",
-                                err = format!("{:?}", e)
+                                err = format!("{:?}", err)
```
Comment (herr-seppia) [inline node/src/chain/fsm.rs:303]:
```suggestion
                                err = format!("{err:?}")
```

File: node/src/chain/fsm.rs [hunk 4]
```diff
@@ -187,134 +190,139 @@ impl<N: Network, DB: database::DB, VM: vm::VMExecution> SimpleFSM<N, DB, VM> {
...
                             );
-                            return Ok(None);
                         }
                     }
+                    Err(e) => {
+                        error!(
+                            event = "revert failed",
+                            err = format!("{:?}", e)
```
Comment (herr-seppia) [inline node/src/chain/fsm.rs:310]:
```suggestion
                            err = format!("{e:?}")
```

File: consensus/src/config.rs [hunk 5]
```diff
@@ -26,11 +26,15 @@ pub const MAX_BLOCK_SIZE: usize = 1_024 * 1_024;
...
 /// Emergency mode is enabled after 16 iterations
 pub const EMERGENCY_MODE_ITERATION_THRESHOLD: u8 = 16;
+pub const EMERGENCY_BLOCK_ITERATION: u8 = u8::MAX;
 
 pub const MIN_STEP_TIMEOUT: Duration = Duration::from_secs(7);
 pub const MAX_STEP_TIMEOUT: Duration = Duration::from_secs(40);
 pub const TIMEOUT_INCREASE: Duration = Duration::from_secs(2);
 
+pub const EMERGENCY_BLOCK_MIN_TIMESTAMP: Duration =
+    Duration::from_secs(MAX_STEP_TIMEOUT.as_secs() * MESSAGE_MAX_ITER as u64);
```
Comment (herr-seppia) [inline consensus/src/config.rs:36]:
```suggestion
    MAX_STEP_TIMEOUT * MESSAGE_MAX_ITER as usize;
```

File: node/src/chain/acceptor.rs (hunk 2 above)
Comment (fed-franz) [reply node/src/chain/acceptor.rs:1359]: Yes. I've added the exception to `verify_prev_block_cert`

File: consensus/src/config.rs (hunk 5 above)
Comment (fed-franz) [reply consensus/src/config.rs]: updated definition

File: node/src/chain/acceptor.rs (hunk 1 above)
Comment (fed-franz) [reply node/src/chain/acceptor.rs]: Done

**Comment by fed-franz** [comment]
This PR has been successfully tested:
- locally, with locally crafted emergency blocks
- on testnet, with no apparent regressions

**Review by herr-seppia — DISMISSED** [review]
LGTM

**Review by herr-seppia — APPROVED** [review]

**Pull request merged**